### DISCUSSION TITLE
Evaluating multiple fact instances of the same type

### DISCUSSION THREAD

**Discussion by eliaslevy** [hyperjumptech/grule-rule-engine#236] [Q&A]
Please correct me if I am wrong, but it appears that, unlike Drools, Grule is incapable of evaluating rules against a set of facts that include multiple instances of the same struct type unless you name each instant something different. But if you name them differently, you can have a rule that will match both instances generically.

[`DataContext.Add()`](https://github.com/hyperjumptech/grule-rule-engine/blob/v1.10.0/ast/DataContext.go#L87-L90) takes key as its first parameter. The fact is stored in map keyed by that first parameter. Any facts added to the`DataContext` using the same key will overwrite previous facts that use the same key. This seem like a severe limitation (e.g. you can't have a set of facts representing order items associated with an order to compute a total, because you can only insert a single order item while a real order may have many items).

If that is the case, it should be clearly mentioned in the documentation.

**Comment by newm4n** [comment]
Grule support facts for array/slice and map. You could construct your slice/array or map and add them into the context. The thing is, there are so many way you can evaluate your fact, process it and be reason with it, element by element or sets by sets. relying on the engine to process/iterate the array element are just not the purpose of the engine. 

Rule engine should tell you WHAT to do, but NOT HOW to do it. You should implement the logic to handle that array of fact, the logic you can put in the context as well as function, let the engine invoke the function that do your specific logic. 

For example, You could create a fact of ShoppingCart that contain an array of items you have in the cart. The ShoppingCart may have functions like `GetTotalWeight() float64`, `GetTotalPrice() float64`, `IsContainsLuxuryItems() boolean`, `ApplyTax(percetage float64)`, etc. The engine just do something like "If the cart contains luxuryItem, add x amount of additional tax percentage", how you know if the cart contains luxury items, and how to apply the tax is supposed to be done by the cart logic. The rule engine only apply some rule to the cart.

**Reply by eliaslevy** [reply]
Yes, you could model your facts a single large complex nested fact, and use helper methods to compute values, but this seems contrary to the spirit of rule engines, which allow you to build up logic from simpler rule statements, and allow the rule developer to encode the business logic in the rule, rather than having the programmer encode it in whatever programming language is used to implement the models.

Again, imagine you are developing a game and want to use the rule engine to define how characters interact with each other. Instead of developing a simple rule that specifies that if a character attacks another character with a sword the victim health is reduced by 10, which could be expressed in Drools with something like:
```
rule "Sword attack"
  when
    Attack( weapon == "Sword", $victim: victim )
  then
    $victim.setHealth( $victim.getHealth() - 10 );
end
```

In Grule such a simple rule is not possible, because you can't have multiple `Attack` objects. Rather, you need to model it as a single `Game` object with an array of `Attacks`, then you must implement in Go a method that iterates over all the attacks searching for the ones that use a sword and update their the victims health. Sure, you could parametrize that method with the weapon to search for and the health to deduct, allowing you to write a rule like:
```
rule SwordAttack "Sword attack" {
  when
    Game.hasAttackType("sword")
  then
    Game.reduceAttackVictimsHealth("sword", 10);
}
```

But such a rule makes almost no use of the rule engine capabilities. `hasAttackType` and `reduceAttackVictimsHealth` are iterating (twice!) over the attacks searching for sword attacks instead of the rule engine's Rete algorithm performing the search. Additionally, the engine doesn't know what victim character's health was updated. We can't really use the `Changed()` function to specify which ones changed.

**Reply by newm4n** [reply]
I can add multiple fact objects into the context. I can add victim, attacker, weapon type into the context and then execute the engine. You do not necessarily create  complex fact contains the whole game entities.

**Reply by eliaslevy** [reply]
@newm4n apologies if I am being dense, but how would you go about doing so? 

If I use `DataContext.Add("Attack", attack)`, the `attack` fact will replace any previous attack that `Attack` referenced, as the reference is stored in a map keyed by the `Attack` string.

I could insert multiple attack facts using different names (e.g. `DataContext.Add("Attack1", attack1)`, `DataContext.Add("Attack2", attack2)`, etc), but then I can't have a single rule that applies to all attack facts.

Did I miss something?

**Reply by newm4n** [reply]
Yes, each rule engine invocation will reflect each of the attack event happening in one specific point in time. If you wish to do multiple attach evaluation, the there should be a fact reference to all of them eg. FirstAttack, SecondAttack, Third... Or have it in an array with helper function in the fact. 

Or, if you still insist to handle multiple attack in a series you can replace the attack fact every time the attack happen and re-run the engine.

```go
ctx.Add("victim", VictimInstance)
ctx.Add("Attacker", AttackerInstance)
for atk := range Attacks {
    ctx.Add("Attack", atk)
    engine.Execute(ctx, kb)
}
```

This mechanism already shown in the example https://github.com/hyperjumptech/grule-rule-engine/blob/master/examples/ItemArrayExample_test.go
